### DISCUSSION TITLE
Highlight every Nth byte withing a specified range or number of times.

### DISCUSSION THREAD

**Discussion by Uzver123** [WerWolv/ImHex#1553] [Get Help]
Look this screenshot i have serialized data `0a`,` 0b`, `0c `etc that repeats every 72 bytes i need to highlight it. 

If i simply highlight all entries that match `0a` this will confuse me as it will highlights `0a` that is also part of other not serialized entries (red box on screenshot).

How can i do it?

![image](https://github.com/WerWolv/ImHex/assets/143356096/c3f98c6a-4b44-412e-aeef-ccee0f778311)

**Comment by WerWolv** [comment]
This can easily be implemented using the pattern language. You can find its documentation on https://docs.werwolv.net

**Reply by Uzver123** [reply]
I was afraid structuring the binary file is the only option, its bit complicated for me and there's not a single tutorial out there on youtube on how to make this templates work in any of hex editor not just imHex.

Could you provide code example related to my original question this is only thing i need, ideally would not want to learn a entire new coding language just to highlight cells in hex editor.

Additionally would be great to have this as a simple and quick option like rule or filter can be applied with several mouse and few key strokes 

- starting offset 
- end offset
- highlight number of bytes 
- repeat every Nth byte

because if someone analyzing unknown binary (without any blueprints available) file structure trying to visualize it create some kind of map by highlighting serialized sections to quickly get idea how data is structured, diving directly into codding and switching mind to how to write code can be distracting from a visualization stage.

How i see it first stage to analyze and then structure and de compile, two different goals that should separate set's of tools.

**Reply by WerWolv** [reply]
How else are you expecting to implement custom logic to find those patterns? Having to write UI for all of this is an enormous amount of work.

If you don't feel like learning the pattern language, you're stuck with doing it manually using e.g Bookmarks.
If you _just_ want to highlight all strings or just want to search for the byte 0x0A, you can use the find view of course but that's also nowhere near as powerful as the pattern language is.

Something like this should work for your usecase. That places a marker byte pattern (for your 0x0A byte) as well as a string pattern every 72 bytes starting from address 0x00 all the way to the end of the file.
```cpp
#include <std/mem.pat>

struct YourStruct {
    u8 marker;
    padding[6];
    char string[];
    padding[72 - ($ - addressof(this))];
};

YourStruct structs[while(!std::mem::eof())] @ 0x00;
```

**Reply by WerWolv** [reply]
ImHex is _really_ not the kind of tool you can just do everything with without at least learning something about how it works. 
Read the docs and looks through the [Pattern Database](https://github.com/WerWolv/ImHex-Patterns) for existing file formats if you want to learn the pattern language, it's really not all that hard.

**Reply by Uzver123** [reply]
That sample code above throws an error.

`E: error[E0004]: Type error.  Array expanded past end of the data.`

Looked for long time at the template for for ZIP file i could not understand it, guess unless you are C++ programming background its going to be tough to learn.

Every little change i try to make to simplify it throws an error. 

Perhaps you need some code samples in the manual that will go trough basics of usage e.g. like on W3Schools they have for Python etc even better educational video that will explain at least basics of usage. 

Right now its a nightmare but so is structuring in any HEX editor its all same C++ based language... 

I will probably go with Kaitai as i see no real benefits of imHex interface and structuring can be done in any other HEX editor.

**Reply by WerWolv** [reply]
Feel free to use whatever you want, just don't expect that everything will just trivially work without you having to learn anything first. Kaitai probably has an even steeper learning curve than ImHex does.

For future reference, `E: error[E0004]: Type error.  Array expanded past end of the data.` means that the array tried to expand past the end of your data, most likely because the code I posted just starts placing these patterns at the start of the file and goes on until it reaches the end of the data, without actually checking anything.
I don't have your file so I can't really help you much more. You'll need to adapt the code to work for your specific usecase.
